### PR TITLE
Issue/4813 site picker recent

### LINKED ISSUE #4813: Site Picker: Sort recently used sites to the top of the site list
In Calypso web, if you have 15 or more sites, there is a "recent sites" group that appears at the top of the site picker (see https://github.com/Automattic/wp-calypso/pull/2742) to keep you from having to scroll every time to find recently used sites.

This shows your most recently used sites at the top of your site list, with other sites shown alphabetically beneath it:

screen shot 2016-11-21 at 4 16 37 pm

We could do the same in the app, showing the three most recently used sites at the top of the site list. Currently, we show the primary site at the top of the list with other sites listed alphabetically beneath that:

h/t Ryan for [raising this issue](https://test.wordpress.com/2016/11/10/session-notes-back-buttons-hidden-sites-touch-image-flow-notifications-site-switching-flow-keyboard-flow-media-section-desktop-first/).

##### Tested on Nexus 9, Android 7.0, WPAndroid alpha-28

Related iOS issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/6217

### PR DESCRIPTION
**Pull request by nbradbury**
Fixes #4813 - Adds a section to the top of the site picker which shows up to the four most recently chosen sites. Only appears when the user has at least 15 unhidden sites, and doesn't appear in search mode.

This should match the UI added in [this framework PR](https://github.com/Automattic/wp-calypso/pull/2742), where the recent sites are shown at the top, followed by a divider, followed by the default site (unless it's also a recent pick).

### CODE REVIEW

**Comment by maxme** [comment]
Looking at Calypso, the "recent sites" come from the "/me/settings" endpoint: https://github.com/Automattic/wp-calypso/blob/4ff89b9d2d9422e0da628a3bfd8fcec1060e93df/client/components/site-selector/index.jsx#L440 (some infos [here](https://github.com/Automattic/wp-calypso/tree/4ff89b9d2d9422e0da628a3bfd8fcec1060e93df/client/state/preferences))

Example call of  `v1.1/me/settings`: 
```json
    "calypso_preferences":
        "recentSites":
            0: 105284648
            1: 108068616
            2: 57991476
```
Should we do the same thing in the apps, if yes, what should we do with self hosted sites?

**Comment by nbradbury** [comment]
@maxme I'm torn on this. Do we really want the recent sites to be the same on web and mobile, or should they reflect the selection the user made on that specific platform? I'm leaning towards the latter. especially when the complications of self-hosted are added to the mix.

**Comment by maxme** [comment]
> @maxme I'm torn on this. Do we really want the recent sites to be the same on web and mobile, or should they reflect the selection the user made on that specific platform? I'm leaning towards the latter, especially when the complications of self-hosted are added to the mix.

Ok that works for me. Can we avoid changing the sites table and store this in AppPrefs instead? That seems more obvious and this will require less work when we merge it in `feature/fluxc-integration` branch.

**Comment by nbradbury** [comment]
> Can we avoid changing the sites table and store this in AppPrefs instead?

Yes, I can make that change.

**Comment by nbradbury** [comment]
@maxme I updated this PR to store recent picks in prefs, so ready for another round.

File: WordPress/src/main/java/org/wordpress/android/models/Blog.java [hunk 1]
```diff
@@ -573,4 +574,12 @@ public boolean hasCapability(Capability capability) {
...
         }
     }
+
+    public void setLastPickedTimestamp(long timestamp) {
+        this.lastPickedTimestamp = timestamp;
+    }
+
+    public long getLastPickedTimestamp() {
+        return lastPickedTimestamp;
+    }
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/models/Blog.java]: We can revert the changes in that file.

File: WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java [hunk 2]
```diff
@@ -394,7 +400,7 @@ protected void onCancelled() {
         @Override
         protected Void doInBackground(Void... params) {
             List<Map<String, Object>> blogs;
-            String[] extraFields = {"isHidden", "dotcomFlag", "homeURL"};
+            String[] extraFields = {"isHidden", "homeURL", "last_picked_timestamp"};
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: This should be reverted as well.

**Comment by nbradbury** [comment]
@maxme Thanks for catching those "timestamp" references - removed in 9d15bde

**Comment by maxme** [comment]
Looks good, :shipit:
